boa/src/builtins/map/mod.rs

```diff
@@ -248,7 +249,31 @@ impl Map {
 
         let size = if let Some(object) = this.as_object() {
             if let Some(map) = object.borrow_mut().as_map_mut() {
-                map.insert(key.clone(), value.clone());
+                let key = match key {
+                    JsValue::Rational(r) => {
+                        if r.is_zero() {
+                            JsValue::Rational(0f64)
+                        } else {
+                            key.into()
+                        }
+                    }
+                    JsValue::Integer(i) => {
+                        if i.is_zero() {
+                            JsValue::Integer(0)
+                        } else {
+                            key.into()
+                        }
+                    }
+                    JsValue::BigInt(b) => {
+                        if b.is_zero() {
+                            JsValue::BigInt(JsBigInt::from(0))
+                        } else {
+                            key.into()
+                        }
+                    }
+                    _ => key.into(),
+                };
```

Comment:
This checks for -0.0f, right? Wouldn't be enough with the `JsValue::Rational` branch?

Reply:
Yep, just realised you're totally right, no need for the extra branches, thanks!